lenskit-eval/src/main/java/org/lenskit/eval/temporal/TemporalEvaluator.java

```diff
@@ -0,0 +1,225 @@
+        rebuildPeriod = unit.toSeconds(time);
+        return this;
+    }
+
+    /**
+     * @param seconds default rebuild period in seconds
+     * @return Itself for  method chaining
+     */
+    public TemporalEvaluator setRebuildPeriod(Long seconds) {
+        rebuildPeriod = seconds;
+        return this;
+    }
+
+    /**
+     * @return Returns prediction output file
+     */
+    public File getPredictOutputFile() {
+        return predictOutputFile;
+    }
+
+    /**
+     * @return Returns rebuild period
+     */
+    public Long getRebuildPeriod() {
+        return rebuildPeriod;
+    }
+
+    /**
+     * During the evaluation, it will replay the ratings, try to predict each one, and
+     * write the prediction, TARMSE and the rating to the output file
+     *
+     * @return Itself for  method chaining
+     */
+
+    public TemporalEvaluator execute() throws IOException, RecommenderBuildException {
+        Preconditions.checkState(algorithm != null, "no algorithm specified");
+        Preconditions.checkState(dataSource != null, "no input data specified");
+        Preconditions.checkState(predictOutputFile != null, "no output file specified");
+
+        TableWriter tableWriter;
+        //Builds file layout
+        TableLayoutBuilder tlb = new TableLayoutBuilder();
+        //file headers
+        tlb.addColumn("User")
+           .addColumn("Item")
+           .addColumn("Rating")
+           .addColumn("Timestamp")
+           .addColumn("Prediction")
+           .addColumn("TARMSE");
+
+        TableLayout tl = tlb.build();
+        tableWriter = CSVWriter.open(predictOutputFile, tl, CompressionMode.AUTO);
+
+        List<Rating> ratings = ObjectStreams.makeList(dataSource.streamEvents(Rating.class, SortOrder.TIMESTAMP));
+        BinaryRatingDAO limitedDao = dataSource.createWindowedView(0);
+        LenskitRecommenderEngine lre;
+        Recommender recommender;
+        double sse = 0;
+        int n = 0;
+        lre = null;
+        long buildTime = 0L;
+        try {
+            for (Rating r : ratings) {
+                if (r.getTimestamp() > 0 && limitedDao.getLimitTimestamp() < r.getTimestamp()) {
+                    limitedDao = dataSource.createWindowedView(r.getTimestamp());
+                }
+                LenskitConfiguration config = new LenskitConfiguration();
+                config.addComponent(limitedDao);
+                if (r.getTimestamp() - buildTime >= rebuildPeriod) {
+                    buildTime = r.getTimestamp();
+                    lre = LenskitRecommenderEngine.newBuilder()
+                                                  .addConfiguration(algorithm.getConfigurations().get(0))
+                                                  .addConfiguration(config, ModelDisposition.EXCLUDED)
+                                                  .build();
+                }
+                recommender = lre.createRecommender(config);
```

Comment:
We can refactor this code so that it is all in the `if`, and so that we only create a recommender once for a block of ratings with the same timestamp. Please do that.

Also, when testing whether we need to build a new engine, we should explicitly test if `lre` is null to always catch the initial condition.